### PR TITLE
desyspider: return also bad harvested data

### PR DESCRIPTION
**Pull request by Glignos**
*Implements logic for returning faulty records
 in a specific format

Signed-off-by: Glignos <glignos93@gmail.com>

## Checklist:

- [ ] I have all the information that I need (if not, move to `RFC` and look for it).
- [ ] I linked the related issue(s) in the corresponding commit logs.
- [ ] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [ ] My code follows the code style of this project.
- [ ] I've added any new docs if API/utils methods were added.
- [ ] I have updated the existing documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: hepcrawl/spiders/desy_spider.py [hunk 1]
```diff
@@ -278,13 +280,21 @@ def _get_marcxml_records(response_body):
...
+                    return {'xml_record': xml_record, 'error': e,
+                            'traceback': traceback.format_tb(sys.exc_info()[2])}
             app = Flask('hepcrawl')
             app.config.update(
                 self.settings.getdict('MARC_TO_HEP_SETTINGS', {})
             )
             with app.app_context():
-                dojson_record = hep.do(object_record)
+                try:
+                    dojson_record = hep.do(object_record)
```
Comment (drjova) [inline hepcrawl/spiders/desy_spider.py]: marcxml2record

File: hepcrawl/spiders/desy_spider.py [hunk 2]
```diff
@@ -278,13 +280,21 @@ def _get_marcxml_records(response_body):
 
     def _hep_records_from_marcxml(self, marcxml_records):
         def _create_json_record(xml_record):
-            object_record = create_record(etree.XML(xml_record))
+            try:
+                object_record = create_record(etree.XML(xml_record))
```
Comment (drjova) [inline hepcrawl/spiders/desy_spider.py]: don't need the ``etree``

File: tests/unit/test_desy.py [hunk 3]
```diff
@@ -114,3 +114,14 @@ def test_pipeline(generated_records, expected_records):
...
+
+def test_faulty_marc():
+    spider = create_spider()
+    path = os.path.abspath(
+        os.path.join('tests/unit/responses/desy/faulty_record.xml', ''))
+    with open(path, 'r') as xmlfile:
+        data = xmlfile.read()
+    result = spider._hep_records_from_marcxml([data])
+    assert type(result[0]['error']) is ValueError
+    assert result[0].get('traceback', None) is not None
```
Comment (drjova) [inline tests/unit/test_desy.py]: ```s/,None//``

File: hepcrawl/spiders/desy_spider.py (hunk 1 above)
Comment (drjova) [reply hepcrawl/spiders/desy_spider.py]: https://github.com/inspirehep/inspire-dojson/blob/f533b083f9993306ae68e025e18eac0eda2e24f3/inspire_dojson/api.py

File: hepcrawl/spiders/desy_spider.py [hunk 4]
```diff
@@ -281,7 +281,7 @@ def _get_marcxml_records(response_body):
     def _hep_records_from_marcxml(self, marcxml_records):
         def _create_json_record(xml_record):
             try:
-                object_record = create_record(etree.XML(xml_record))
+                object_record = create_record(xml_record)
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py]: you don't need `create_record`. `marcxml2record` takes a string containing XML and spits out, if all goes well, a JSON record.

File: hepcrawl/spiders/desy_spider.py [hunk 5]
```diff
@@ -10,10 +10,12 @@
 from __future__ import absolute_import, division, print_function
 
 import os
+import sys
+import traceback
 
 from dojson.contrib.marc21.utils import create_record
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py]: no need to import this now.

File: hepcrawl/spiders/desy_spider.py [hunk 6]
```diff
@@ -278,17 +280,20 @@ def _get_marcxml_records(response_body):
...
             with app.app_context():
-                dojson_record = hep.do(object_record)
+                try:
+                    hep_record = marcxml2record(xml_record)
+                except Exception as e:
+                    return {'xml_record': xml_record, 'error': e,
+                            'traceback': traceback.format_tb(sys.exc_info()[2])}
                 base_uri = self.settings['SCHEMA_BASE_URI']
-                dojson_record['$schema'] = base_uri + 'hep.json'
+                hep_record['$schema'] = base_uri + 'hep.json'
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py]: this is not needed if validation is not done in hepcrawl any more. You may need to synchronize with @iulianav about who removes this.

File: tests/unit/test_desy.py [hunk 7]
```diff
@@ -114,3 +114,14 @@ def test_pipeline(generated_records, expected_records):
     sorted_generated_records = deep_sort(clean_generated_records)
     sorted_expected_records = deep_sort(expected_records)
     assert sorted_generated_records == sorted_expected_records
+
+
+def test_faulty_marc():
+    spider = create_spider()
+    path = os.path.abspath(
+        os.path.join('tests/unit/responses/desy/faulty_record.xml', ''))
```
Comment (drjova) [inline tests/unit/test_desy.py]: do we need the join? :)

**Pull request merged**